### PR TITLE
authz: implement `PermsFetcher` interface for GitLab authz providers

### PR DESCRIPTION
**Pull request by unknwon**
This PR implements `PermsFetcher` interface for GitLab authz providers, both `SudoProvider` and `OAuthProvider`.

### Notes
- Tests will be added in followup PRs once I run through the new permissions syncing process end-to-end.
- This PR does not change any functionality currently in use, safe to merge upon approvals.
- This PR explicitly mentions in `PermsFetcher` definition that implementations should return "partial but valid results in case of error". The rationales are:
	- API calls to code host are expensive.
	- When a new user or repo just being added, there is literally no permissions, something (valid) is better than nothing. LMK any objections, if anyone thinks this will make debugging harder.

Implements #8492.

### CODE REVIEW

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	client := p.clientProvider.GetOAuthClient(tok.AccessToken)
+	return listProjects(ctx, client)
+}
+
+// FetchRepoPerms is a stub implementation for OAuth authz provider because the API
+// endpoint we use requires admin in order to get complete results.
+func (p *OAuthAuthzProvider) FetchRepoPerms(context.Context, *api.ExternalRepoSpec) ([]string, error) {
+	return []string{}, nil
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]: Return nothing here for OAuth authz provider is doing any worse because it means:
1. We can't do proactive syncing in repo-centric
2. We can still do proactive syncing for users created in the database (right after user sign up), which is still better than our current on-demand syncing.

File: internal/extsvc/gitlab/members.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/peterhellberg/link"
+)
+
+// Member contains fields for a group or project membership.
+type Member struct {
```
Comment (unknwon) [inline internal/extsvc/gitlab/members.go]: On a glance, this struct is similar to `type User struct` we already have, but they have different meaning and response struct from GitLab APIs. I do not think mixing them together is a good idea while some fields are guaranteed to not exist in another struct, and vice versa.

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/gitlab"
+)
+
+// FetchUserPerms returns a list of project IDs (on code host) that the given account
+// has read access on the code host. The project ID has the same value as it would be
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]:
```suggestion
// has read access to. The project ID has the same value as it would be
```

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/gitlab"
+)
+
+// FetchUserPerms returns a list of project IDs (on code host) that the given account
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]:
```suggestion
// FetchUserPerms returns a list of private project IDs (on code host) that the given account
```

I would move this here because it's an important if I understood it correctly: the list doesn't contain public projects, yes?

File: enterprise/cmd/frontend/internal/authz/gitlab/sudo.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "get external account data")
+	}
+
+	client := p.clientProvider.GetPATClient(p.sudoToken, strconv.Itoa(int(user.ID)))
+	return listProjects(ctx, client)
+}
+
+// listProjects is a helper function to request for all private projects that are accessible
+// (20 => Reporter access) by the authenticated or impersonated user in the client. It may
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/authz/gitlab/sudo.go]:
```suggestion
// (access level: 20 => Reporter access) by the authenticated or impersonated user in the client. It may
```

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go (hunk 1 above)
Comment (keegancsmith) [reply enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]: won't an empty slice lead to us removing all users who have had access granted via the FetchUserPerms call?

File: enterprise/cmd/repo-updater/authz/perms_syncer.go [hunk 6]
```diff
@@ -205,6 +216,9 @@ func (s *PermsSyncer) syncRepoPerms(ctx context.Context, repoID api.RepoID) erro
 	if err != nil {
 		return errors.Wrap(err, "fetch repository permissions")
 	}
+	if len(usernames) == 0 {
+		return nil
+	}
```
Comment (keegancsmith) [inline enterprise/cmd/repo-updater/authz/perms_syncer.go]: I think signalling an empty list as do no work can lead to weird bugs. IE I can imagine a repo which has no users that are allowed to access it. EG a private repo can have some users, then an admin removes all users to lock it down. With this code path it looks like we would never update the perms on our side?

Comment (unknwon) [reply enterprise/cmd/repo-updater/authz/perms_syncer.go]: Good point, I'll come up a better handling to signal caller "nothing happened" or "not supported".

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+// used as api.ExternalRepoSpec.ID. The returned list only includes private project IDs.
+//
+// This method may return partial but valid results in case of error, and it is up to
+// callers to decide whether to discard.
+//
+// API docs: https://docs.gitlab.com/ee/api/projects.html#list-all-projects
+func (p *OAuthAuthzProvider) FetchUserPerms(ctx context.Context, account *extsvc.ExternalAccount) ([]string, error) {
+	if account == nil {
+		return nil, errors.New("no account provided")
+	} else if account.ServiceType != p.codeHost.ServiceType || account.ServiceID != p.codeHost.ServiceID {
```
Comment (tsenart) [inline enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]: Can we factor this out into something like https://github.com/sourcegraph/sourcegraph/blob/f62b33a59e099941bfb72d6e22647dcc891e06fe/internal/extsvc/codehost.go#L38?

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "get external account data")
+	}
+
+	client := p.clientProvider.GetOAuthClient(tok.AccessToken)
+	return listProjects(ctx, client)
+}
+
+// FetchRepoPerms is a stub implementation for OAuth authz provider because the API
+// endpoint we use requires admin in order to get complete results.
```
Comment (tsenart) [inline enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]: Don't we have an admin access token in the GitLab external service?

File: enterprise/cmd/frontend/internal/authz/gitlab/sudo.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+// listProjects is a helper function to request for all private projects that are accessible
+// (access level: 20 => Reporter access) by the authenticated or impersonated user in the client. It may
+// return partial but valid results in case of error, and it is up to callers to decide
+// whether to discard.
+func listProjects(ctx context.Context, client *gitlab.Client) ([]string, error) {
+	q := make(url.Values)
+	q.Add("visibility", "private")  // This method is meant to return only private projects
+	q.Add("min_access_level", "20") // 20 => Reporter access (i.e. have access to project code)
+	q.Add("per_page", "100")        // 100 is the maximum page size
```
Comment (tsenart) [inline enterprise/cmd/frontend/internal/authz/gitlab/sudo.go]: Is page size configurable in GitLab? If so, a larger page size could really speed things up. If it's possible, we should ask admins to do it and tell Sourcegraph the page size via external service config (or similar).

File: enterprise/cmd/frontend/internal/authz/gitlab/sudo.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	return projectIDs, nil
+}
+
+// FetchRepoPerms returns a list of user IDs (on code host) who have read ccess to
+// the given project on the code host. The user ID has the same value as it would
+// be used as extsvc.ExternalAccount.AccountID. The returned list includes both
+// direct access and inherited from the group membership.
+//
+// API docs: https://docs.gitlab.com/ee/api/members.html#list-all-members-of-a-group-or-project-including-inherited-members
+func (p *SudoProvider) FetchRepoPerms(ctx context.Context, repo *api.ExternalRepoSpec) ([]string, error) {
```
Comment (tsenart) [inline enterprise/cmd/frontend/internal/authz/gitlab/sudo.go]: It would be nice that both `FetchRepoPerms` and `FetchUserPerms` would have a more descriptive return type. Rather than a bare `string` could we have some type aliases? Example: `type ExternalUserID string` and `type ExternalRepoID string`

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go (hunk 7 above)
Comment (unknwon) [reply enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]: Nice! Will do.

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go (hunk 8 above)
Comment (unknwon) [reply enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]: Ah! I was wondering why we also have a PERSONAL_ACCESS_TOKEN for [GitLab OAuth application](https://docs.sourcegraph.com/admin/repo/permissions#oauth-application) in our docs.

File: enterprise/cmd/frontend/internal/authz/gitlab/sudo.go (hunk 9 above)
Comment (unknwon) [reply enterprise/cmd/frontend/internal/authz/gitlab/sudo.go]: The top result shows there is no resolution to this 😞 https://gitlab.com/gitlab-org/gitlab-foss/-/issues/28675

TL;DR The GitLab team believes this opens the door to reduce overall GitLab performance in unexpected ways.

File: enterprise/cmd/frontend/internal/authz/gitlab/oauth.go (hunk 7 above)
Comment (unknwon) [reply enterprise/cmd/frontend/internal/authz/gitlab/oauth.go]: Oh haha, GoLand tells me this function is actually unused 😂 Poor man.

**Comment by unknwon** [comment]
Add feedbacks addressed, please take another look! @keegancsmith @tsenart
